teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -26,54 +46,387 @@ This document proposes updating Tekton Pipelines' feature flags design, as origi
+
+### Per-feature flag for _new_ API-driven features
+Introduce per-feature flags for each **new** API-driven feature. Each feature will have its own flag, instead of using the group API-driven flag `enable-api-fields` to enable or disable all features of a stability level. Note that our proposal only takes effect on api-driven features while behavioural flags will remain the existing behaviour.
+
+The flag will also include release information and stability level for the feature as the source of truth. New behavioural features will also have a new per-feature flag to either enable or disable the feature. This will allow behavioural features that have values leading to behaviors in different stability levels to be turned on or off instead of depending on the stability level of features that are enabled.
+
+For example:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: feature-flags
+data:
+  # alpha: v0.53
+  #
+  # Pipeline in pipeline has ... functionalities.
+  # It is disabled by default.
+  enable-pipeline-in-pipeline: "false"
+  # alpha: v0.53
+  # 
+  # Trusted artifacts has ... functionalities.
+  # It is disabled by default. Set it to true to enable this feature.
+  enable-trusted-artifacts: "true"
+```
+
+See [implementation plan](#implementation-plan) for more details on the PerFeatureFlag struct.
+
+All **new** features can only be enabled via per-feature flags. When they first get introduced as `alpha`, they will be disabled by default. When new features get promoted to `stable`, they will be enabled by default according to the following table:
+
+| Feature stability level | Default  |
+| ----------------------- | -------- |
+| Stable                  | Enabled; Cannot be disabled  |
+| Beta                    | Disabled |
+| Alpha                   | Disabled |
+
+Note that per-feature flags that have stabilized cannot be disabled. We will deprecate the per-feature flag after it has become stable and then remove it eventually after the deprecation period according to the API compatibility policy. We will give deprecation and later removel notice of the per-feature flags via release notes after their promotion to `stable`. Cluster operators who do not want such opt-in features would have enough notice to implement admission controllers on their own to disable the feature.
```

Comment:
@JeromeJu sorry, I thought I commented on this before, but I must have missed it 🙏 

Keeping the feature flag around, deprecated but unusable does not make much sense to me.
Making it unusable provides the same behaviour as removing it, the deprecation has no effect really.

The only reason I could see to keep the flag around would be if we started validating the content of the config file (which we do not do today) - but even so, I think failing config validation would be preferrable to silently ignoring a feature flag.

Reply:
Never mind, I think I misread this section 🙏
A small update to the table might help  - but feel free to ignore this - I can also propose a PR with the update if that's ok for you.

```
| Feature stability level | Default  |
| ----------------------- | -------- |
| Stable                  | Enabled; Cannot be disabled once the flag is removed (after deprecation) |
| Beta                    | Disabled |
| Alpha                   | Disabled |
```

Reply:
Thanks for the suggestion, PTAL https://github.com/tektoncd/community/pull/1091 🙏